### PR TITLE
build: Add race detector flag

### PR DESCRIPTION
**Pull request by javierhonduco**
Let's add [Golang's race detector](https://go.dev/doc/articles/race_detector) to our Makefile. I have been running it in my test environment, and it showed zero errors, even under a lot of pressure (lots of concurrent debug info uploads, etc).

With the `GORACE="log_path=/tmp/race"` env variable we can pick a path where to write its output, rather than stdout / stderr.

Right now we should run this locally, but once https://github.com/parca-dev/parca-agent/issues/463 lands, this will be run in CI, too! 🎉 

Fixes: https://github.com/parca-dev/parca-agent/issues/445
## Test plan

**Build without sanitisers works**:

```
[javierhonduco@computer parca-agent]$ make build-dyn 
GOOS=linux GOARCH=amd64 CC=clang CGO_CFLAGS="-I /home/javierhonduco/code/parca-agent/dist/libbpf/usr/include" CGO_LDFLAGS="/home/javierhonduco/code/parca-agent/dist/libbpf/libbpf.a" go build  -tags osusergo,netgo -trimpath -v -o dist/parca-agent ./cmd/parca-agent
```

**Build with ASAN works**:

```
[javierhonduco@computer parca-agent]$ make build-dyn ENABLE_ASAN=yes
GOOS=linux GOARCH=amd64 CC=clang CGO_CFLAGS="-I /home/javierhonduco/code/parca-agent/dist/libbpf/usr/include" CGO_LDFLAGS="/home/javierhonduco/code/parca-agent/dist/libbpf/libbpf.a" go build -asan -tags osusergo,netgo -trimpath -v -o dist/parca-agent ./cmd/parca-agent
```

**Build with race detector works**:

```
[javierhonduco@computer parca-agent]$ make build-dyn ENABLE_RACE=yes
GOOS=linux GOARCH=amd64 CC=clang CGO_CFLAGS="-I /home/javierhonduco/code/parca-agent/dist/libbpf/usr/include" CGO_LDFLAGS="/home/javierhonduco/code/parca-agent/dist/libbpf/libbpf.a" go build -race -tags osusergo,netgo -trimpath -v -o dist/parca-agent ./cmd/parca-agent
```

_note_ ASAN + race isn't supported yet. I don't think we should check for this in the Makefile as it will increase its complexity and this is something that might change in the future:

```
[javierhonduco@computer parca-agent]$ make build-dyn ENABLE_ASAN=yes ENABLE_RACE=yes
GOOS=linux GOARCH=amd64 CC=clang CGO_CFLAGS="-I /home/javierhonduco/code/parca-agent/dist/libbpf/usr/include" CGO_LDFLAGS="/home/javierhonduco/code/parca-agent/dist/libbpf/libbpf.a" go build -asan -race -tags osusergo,netgo -trimpath -v -o dist/parca-agent ./cmd/parca-agent
go: may not use -race and -asan simultaneously
```

(`make test ENABLE_RACE=yes` and `make test ENABLE_ASAN=yes` also worked without issues)

### CODE REVIEW

**Review by kakkoyun — COMMENTED** [review]
LGTM.

What does prevent us to utilize this in CI right now?

**Comment by kakkoyun** [comment]
> Nothing that I can think of, we would have to add another job with the right sanitiser enabled and some extra options (such as crashing in case of a race condition instead of continuing, perhaps).
> 
> Adding them in these tests would be a good first step, but enabling them in the integration tests will pay way higher dividends!

Let's do it as part of this PR! @Sylfrena is already working on the e2e CI action. She can make sure necessary make is called properly.

https://github.com/parca-dev/parca-agent/blob/ba3b7410f79a7820c58b9b66c200a5bbe82474b5/Makefile#L259-L262

**Comment by javierhonduco** [comment]
Mm, these are passing locally for me. The only difference I can see is that my patch version is earlier than the one in CI, will try with the same version we run in CI

File: .github/workflows/build.yml [hunk 1]
```diff
@@ -135,6 +135,9 @@ jobs:
       - name: Test
         run: make test
```
Comment (kakkoyun) [inline .github/workflows/build.yml:136]: We can get rid of this one.

**Pull request merged**